### PR TITLE
[12.1.x] ISPN-13116 SFS failures on upgrade from 11.0.9.Final -> 12.1.3.Final

### PR DESCRIPTION
**Pull request by ryanemerson**
https://issues.redhat.com/browse/ISPN-13116
https://issues.redhat.com/browse/ISPN-13128

### CODE REVIEW

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 1]
```diff
@@ -215,14 +227,93 @@ public boolean isAvailable() {
...
+            FileEntry fe = new FileEntry(oldFilePos, buf);
+            oldFilePos += KEY_POS_LATEST;
+
+            // Read old entry content and then write
+            ByRef.Long offset = new ByRef.Long(oldFilePos);
+            K key = unmarshallObject(buf, offset, fe.keyLen);
+            V value = unmarshallObject(buf, offset, fe.dataLen);
+
+            int metaLen = fe.metadataLen > 0 ? fe.metadataLen - TIMESTAMP_BYTES : 0;
+            Metadata metadata = unmarshallObject(buf, offset, metaLen);
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:279]: @pruivo Unmarshalling the entries actually made it more readable and easier to debug :+1:

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 2]
```diff
@@ -255,17 +346,46 @@ private void rebuildIndex() throws Exception {
...
+      File newFile = new File(file.getParentFile(), cacheName + "_new.dat");
+      if (newFile.exists()) {
+         newFile.delete();
+      }
+      try (FileChannel newChannel = SecurityActions.openFileChannel(newFile)) {
+         // Write latest Magic and copy old content
+         newChannel.truncate(0);
+         newChannel.write(ByteBuffer.wrap(MAGIC_LATEST), 0);
+         channel.position(MAGIC_12_0.length);
+         newChannel.transferFrom(channel, MAGIC_12_0.length, channel.size());
```
Comment (wburns) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:361]: Should this be `newChannel.transferFrom(channel, MAGIC_12_0.length, channel.size() - MAGIC_12_0.length);` ? Maybe it just ignores if count is larger than the amount of available bytes?

Ah yeah it does `long max = Math.min(count, src.size() - pos);`

Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:361]: > Maybe it just ignores if count is larger than the amount of available bytes?

Yeah the docs say the following:

```qoute
Fewer than the requested number of bytes will be transferred if the source channel has fewer than count bytes remaining, or if the source channel is non-blocking and has fewer than count bytes immediately available in its input buffer.
```

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 3]
```diff
@@ -361,37 +454,35 @@ private void migrateFromV11() {
...
-      // Read using raw user marshaller without MarshallUserObject wrapping
-      byte[] bytes = buf.array();
-      Marshaller marshaller = ctx.getPersistenceMarshaller().getUserMarshaller();
-      try {
-         return (T) marshaller.objectFromByteBuffer(bytes, 0, length);
-      } catch (IllegalArgumentException e) {
-         // For internal cache key/values and custom metadata we need to use the persistence marshaller
-         return (T) ctx.getPersistenceMarshaller().objectFromByteBuffer(bytes, 0, length);
+      buf.flip();
+      byte[] bytes = new byte[buf.remaining()];
```
Comment (wburns) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:471]: Is there a reason we can't use the underlying byte[] anymore? As far as I can see the underlying byte[] isn't retained in the methods it is passed to.

File: core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+   @DataProvider(name = "testFiles")
+   Object[][] singleTypes() {
+      return new Object[][]{
+            // Not possible to test corrupt Marshaller.JAVA_SERIALIZATION bytes as previously 11.x -> 12. migration always fails with non-protostream marshallers
+            // MAGIC_11_0
+            {"sfs/11_0/all-entries-java-serialization.dat", Marshaller.JAVA_SERIALIZATION, 1003},
+            // MAGIC_12_0 created on 12.1.x
+            {"sfs/12_0/all-entries-java-serialization.dat", Marshaller.JAVA_SERIALIZATION, 1003},
+            // 11.x -> 12.1.x migration resulting in corrupt data
+            {"sfs/corrupt/all-entries.dat", Marshaller.PROTOSTREAM, 2003},
```
Comment (wburns) [inline core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java:121]: Why do we have an extra 1000 entries? Did the corruption create new ones?

File: core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      Files.copy(is, Paths.get(tmpDirectory).resolve(CACHE_NAME + ".dat"), StandardCopyOption.REPLACE_EXISTING);
+
+      ConfigurationBuilderHolder cbh = new ConfigurationBuilderHolder();
+      if (marshallerType == Marshaller.JAVA_SERIALIZATION)
+         cbh.getGlobalConfigurationBuilder().serialization().marshaller(new JavaSerializationMarshaller());
+
+      cbh.newConfigurationBuilder(CACHE_NAME)
+            .persistence()
+            .addSingleFileStore()
+            .segmented(false)
```
Comment (wburns) [inline core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java:139]: I totally forgot about this before, but what about if segmented was set to true? I *think* it might be okay since the way segmentation worked it just created N SingleFileStore instances, so they should all migrate independently, but I am not 100% on that.

File: core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java (hunk 4 above)
Comment (ryanemerson) [reply core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java:121]: It's due to recreating the lost `created` and `lastUsed` timestamps. In the other stores the values have expired so they aren't iterated over.

File: core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java (hunk 5 above)
Comment (ryanemerson) [reply core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java:139]: That's how it happens. The original operator issues that uncovered this issue show that you get a migration log message for each segment with values in SFS.

File: core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java (hunk 4 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java:121]: Makes sense. Is there a place we should mention that expired entries may be resurrected when hitting this migration issue somewhere?

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java (hunk 3 above)
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:471]: Tbh I wasn't sure how correct/safe the original approach was, hence the change. I can revert this though if it's not going to cause any issues.

File: core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java (hunk 4 above)
Comment (ryanemerson) [reply core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java:121]: I'm going to create a small doc describing the original problem as well as the various limitations of the recovery. Downstream we will need this for support. I guess I could put a condensed version of this in the upgrade notes?

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java (hunk 3 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:471]: Yeah, it should be safe. The `objectFromByteBuffer` methods shouldn't retain a reference so we should be fine just using the underlying byte[] in the `ByteBuffer` and can avoid the extra byte[] allocation.

File: core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java (hunk 4 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java:121]: Yeah, I am not sure exactly where it should go upstream. But I guess that is as good as a spot as any. Just make sure to list the previously affected versions explicitly.

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 6]
```diff
@@ -215,14 +227,93 @@ public boolean isAvailable() {
...
+
+            int metaLen = fe.metadataLen > 0 ? fe.metadataLen - TIMESTAMP_BYTES : 0;
+            Metadata metadata = unmarshallObject(buf, offset, metaLen);
+            // internalMetadata was ALWAYS lost - so don't read it
+
+            oldFilePos = offset.get();
+
+            // LastUsed and Created timestamps were lost, set creation time to current time if fe.expiryTime has been set
+            long currentTS = fe.expiryTime > 0 ? timeService.wallClockTime() : -1;
+            MarshallableEntry<? extends K, ? extends V> me = (MarshallableEntry<? extends K, ? extends V>) ctx.getMarshallableEntryFactory().create(key, value, metadata, null, currentTS, currentTS);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:286]: I'm not sure about using `null` for internal metadata. I know we will have issues with xsite because the versions are lost but I'm not sure if `null` is ok for optimistic transactions :/

Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:286]: Right, I forgot about initializing these values. Based upon previous conversations, WDYT to the following approach:

Optimistic tx -> `PrivateMetadata` with version == 1

XSite:

- Attempt to recover latest version from global state when shutdown
- Advise performing xsite state transfer once corrupt data migrated/

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java (hunk 3 above)
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java]: Updated to just use `byte[] bytes = buf.array();`

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java (hunk 6 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:286]: looks good.

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 7]
```diff
@@ -255,17 +297,128 @@ private void rebuildIndex() throws Exception {
...
+   // Initialise missing internal metadata state for corrupt data
+   private PrivateMetadata generateMissingInternalMetadata(K key) {
+      // Optimistic Transactions
+      AdvancedCache<?, ?> cache = ctx.getCache().getAdvancedCache();
+      ComponentRegistry componentRegistry = cache.getComponentRegistry();
+      Configuration config = cache.getCacheConfiguration();
+      TransactionConfiguration txConfig = config.transaction();
+      PrivateMetadata.Builder builder = new PrivateMetadata.Builder();
+      if (txConfig.transactionMode() == TransactionMode.TRANSACTIONAL && txConfig.lockingMode() == LockingMode.OPTIMISTIC) {
+         builder.entryVersion(new NumericVersionGenerator().generateNew());
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:309]:
```suggestion
         builder.entryVersion(new SimpleClusteredVersion(1, 1));
```

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 8]
```diff
@@ -255,17 +297,128 @@ private void rebuildIndex() throws Exception {
...
+      }
+
+      // Async XSite
+      if (config.sites().hasAsyncEnabledBackups()) {
+         KeyPartitioner keyPartitioner = componentRegistry.getComponent(KeyPartitioner.class);
+         IracVersionGenerator ivg = componentRegistry.getIracVersionGenerator().running();
+         int segment = keyPartitioner.getSegment(key);
+         builder.iracMetadata(ivg.generateMetadataWithCurrentVersion(segment));
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:317]:
```suggestion
         String siteName = cache.getCacheManager().getTransport().localSiteName();
         IracEntryVersion version = new IracEntryVersion(Collections.singletonMap(siteName, TopologyIracVersion.newVersion(1)));
         IracMetadata metadata = new IracMetadata(siteName, version);
         builder.iracMetadata(metadata);
```

Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:317]: Isn't this bypassing the stored global state?

Comment (pruivo) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:317]: yes. unfortunately, I don't see a way of breaking the cyclic dependency 
We need to document this. The user must trigger the state transfer both ways; this will trigger a conflict and the merge policy should fix it and assign the same version on both sides for each key.

**Review by wburns — APPROVED** [review]
LGTM for the pieces I commented on.

File: core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+            .location(tmpDirectory);
+
+      try (EmbeddedCacheManager cacheManager = new DefaultCacheManager(cbh, true)) {
+         Cache<Object, Object> cache = cacheManager.getCache(CACHE_NAME);
+         ByRef.Integer i = new ByRef.Integer(0);
+         // Iterate all entries to ensure values can be read
+         cache.forEach((k, v) -> i.inc());
+         assertEquals(numEntries, i.get());
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java:147]: I would add 2 writes here. 
* write to a migrated key
* write to a new key

below, check if both are still there too :)

File: core/src/test/java/org/infinispan/persistence/file/SingleFileStorePopulator.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+//import org.infinispan.persistence.spi.MarshallableEntryFactory;
+//import org.infinispan.test.TestingUtil;
+//
+///**
+// * // TODO: Document this
+// *
+// * @author Ryan Emerson
+// * @since 13.0
+// */
+//public class SingleFileStorePopulator {
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/persistence/file/SingleFileStorePopulator.java:20]: maybe move this class to tools module?

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 11]
```diff
@@ -255,17 +294,182 @@ private void rebuildIndex() throws Exception {
...
+      }
+
+      // Data is not corrupt, so simply update magic and transfer file contents
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startMigratingPersistenceData(cacheName);
+      File newFile = new File(file.getParentFile(), cacheName + "_new.dat");
+      if (newFile.exists()) {
+         newFile.delete();
+      }
+      try (FileChannel newChannel = SecurityActions.openFileChannel(newFile)) {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:334]: I guess if the "magic value" has the same length, you can update the old file directly instead of copying everything to a new file.

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 12]
```diff
@@ -255,17 +294,182 @@ private void rebuildIndex() throws Exception {
...
+      }
+      rebuildIndex();
+   }
+
+   private void migrateCorruptDataV12_0() {
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startRecoveringCorruptPersistenceData(cacheName);
+      File newFile = new File(file.getParentFile(), cacheName + "_new.dat");
+
+      // Dey before the eelease of Infinispan 10.0.0.Final
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:363]:
```suggestion
      // Day before the release of Infinispan 10.0.0.Final
```

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 13]
```diff
@@ -255,17 +294,182 @@ private void rebuildIndex() throws Exception {
...
+            .toEpochMilli();
+      long currentTs = timeService.wallClockTime();
+
+      ByteBuffer buf = ByteBuffer.allocate(KEY_POS_LATEST);
+      try (FileChannel newChannel = SecurityActions.openFileChannel(newFile)) {
+         //Write Magic
+         newChannel.truncate(0);
+         newChannel.write(ByteBuffer.wrap(MAGIC_LATEST), 0);
+
+         long oldFilePos = MAGIC_LATEST.length;
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:378]:
```suggestion
         long oldFilePos = MAGIC_12_0.length;
```

File: core/src/test/java/org/infinispan/persistence/file/SingleFileStorePopulator.java (hunk 10 above)
Comment (ryanemerson) [reply core/src/test/java/org/infinispan/persistence/file/SingleFileStorePopulator.java]: I've removed the file. The content of it is in the `SingleFileGracefulShutdownMigrationTest` JavaDocs.

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 14]
```diff
@@ -255,17 +277,159 @@ private void rebuildIndex() throws Exception {
...
+      }
+      rebuildIndex();
+   }
+
+   private void migrateCorruptDataV12_0() {
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startRecoveringCorruptPersistenceData(cacheName);
+      File newFile = new File(file.getParentFile(), cacheName + "_new.dat");
+
+      // Day before the eelease of Infinispan 10.0.0.Final
```
Comment (wburns) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:323]: release

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 15]
```diff
@@ -255,17 +277,159 @@ private void rebuildIndex() throws Exception {
...
+
+            buf.flip();
+            FileEntry fe = new FileEntry(oldFilePos, buf);
+            // Semantic check to find valid FileEntry
+            if (fe.size <= 0 || fe.keyLen < 0 || fe.dataLen <= 0 || fe.metadataLen < 0 || fe.internalMetadataLen < 0 || fe.expiryTime < -1) {
+               // Check failed, try to read FileEntry from the next byte
+               oldFilePos++;
+               continue;
+            }
+
```
Comment (wburns) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:353]:
```java
If (fe.keyLen == 0) {
  // Just to optimize the previously removed key byte forward search a bit
  oldFilePos += fe.dataLen + fe.metadataLen;
}
```
Since the `unmarshallObject` method below is guaranteed to fail and we know we can skip this amount at least (since fe.size is broken).

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 16]
```diff
@@ -255,17 +277,159 @@ private void rebuildIndex() throws Exception {
...
+               created = buf.getLong();
+               lastUsed = buf.getLong();
+
+
+               // If the Timestamps are in the future or < sanityEpoch, then we're migrating corrupt data so set the value to current wallClockTime
+               if (created != -1 && (created > currentTs || created < sanityEpoch))
+                  created = currentTs;
+
+               if (lastUsed != -1 && (lastUsed > currentTs || lastUsed < sanityEpoch))
+                  created = currentTs;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:388]: `lastUsed = currentTs;`

File: persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java [hunk 17]
```diff
@@ -162,7 +162,7 @@ public void createMetaTable(Connection conn) throws PersistenceException {
    @Override
    public void updateMetaTable(Connection conn) throws PersistenceException {
       short version = Version.getVersionShort();
-      int segments = ctx.getCache().getCacheConfiguration().clustering().hash().numSegments();
+      int segments = ctx.getConfiguration().segmented() ? ctx.getCache().getCacheConfiguration().clustering().hash().numSegments() : -1;
```
Comment (wburns) [inline persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java:165]: Seems this JDBC stuff should be a different PR?

Comment (ryanemerson) [reply persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java:165]: It was uncovered by the changes to `PersistenceCompatibilityTest`. I think it needs to be part of this PR as this could cause issues in the operator upgrade case in a similar way to the SFS issues.

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java (hunk 15 above)
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:353]: It's probably unlikely, but isn't it possible that we get a false positive reading `FileEntry` that has `fe.keyLen=0` with other garbage values for `fe.dataLen` etc? The current code is less efficient, but it will guard against this.

Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:349]: I think moving to the next byte is the only safe way to handle this, e.g:

```java
If (fe.keyLen == 0) {
  oldFilePos++;
  continue;
}
```

Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:349]: Agreed. Forgot that the size could be 0 and the other unmarshalling validations can't be checked.

Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:349]: In which case just change if above to be `fe.keyLen <= 0`. Just put a comment that technically 0 is valid, but means we should be skipping the entry.

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 18]
```diff
@@ -255,17 +273,159 @@ private void rebuildIndex() throws Exception {
...
+            V value;
+            Metadata metadata;
+            ByRef.Long offset = new ByRef.Long(oldFilePos + KEY_POS_LATEST);
+            try {
+               // Read old entry content and then write
+               key = unmarshallObject(buf, offset, fe.keyLen);
+               value = unmarshallObject(buf, offset, fe.dataLen);
+
+               int metaLen = fe.metadataLen > 0 ? fe.metadataLen - TIMESTAMP_BYTES : 0;
+               metadata = unmarshallObject(buf, offset, metaLen);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:360]: I think you should check if metaLen > 0 before trying to unmarshall the Metadata

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 19]
```diff
@@ -255,17 +273,162 @@ private void rebuildIndex() throws Exception {
       }
    }
 
+   // Initialise missing internal metadata state for corrupt data
+   private PrivateMetadata generateMissingInternalMetadata() {
+      // Optimistic Transactions
+      AdvancedCache<?, ?> cache = ctx.getCache().getAdvancedCache();
+      Configuration config = cache.getCacheConfiguration();
+      TransactionConfiguration txConfig = config.transaction();
+      PrivateMetadata.Builder builder = new PrivateMetadata.Builder();
```
Comment (wburns) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:282]: We should probably only initialize this builder if one of the ifs below are true. Otherwise all entries will have an empty internal metadata taking up space it doesn't need to.

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 20]
```diff
@@ -255,17 +273,162 @@ private void rebuildIndex() throws Exception {
...
+            buf = readChannel(buf, oldFilePos, KEY_POS_LATEST);
+            // EOF reached
+            if (buf.remaining() > 0)
+               break;
+
+            buf.flip();
+            FileEntry fe = new FileEntry(oldFilePos, buf);
+            // Semantic check to find valid FileEntry
+            // fe.keyLen = 0 is valid, however it means we should skip the entry
+            if (fe.size <= 0 || fe.keyLen <= 0 || fe.dataLen <= 0 || fe.metadataLen < 0 || fe.internalMetadataLen < 0 || fe.expiryTime < -1) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:345]: Can we add something like `fe.size < fe.keyLen + fe.dataLen + fe.metadataLen + fe.internalMetadataLen` as well? Just to eliminate waiting for a throwable to determine it was corrupted

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 21]
```diff
@@ -255,17 +273,162 @@ private void rebuildIndex() throws Exception {
...
+               // Read old entry content and then write
+               key = unmarshallObject(buf, offset, fe.keyLen);
+               value = unmarshallObject(buf, offset, fe.dataLen);
+
+               int metaLen = fe.metadataLen > 0 ? fe.metadataLen - TIMESTAMP_BYTES : 0;
+               if (metaLen > 0)
+                  metadata = unmarshallObject(buf, offset, metaLen);
+
+               // Entries successfully unmarshalled so it's safe to increment oldFilePos to FileEntry+offset so brute-force can resume on next iteration
+               oldFilePos = offset.get();
```
Comment (wburns) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:365]: Technically this line and the same exact one in the if below can be removed and put the exact line after the if. But I am fine leaving it how it is.

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java (hunk 19 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:282]: this is only invoked when `fe.internalMetadataLen > 0`. 
It is never empty :)

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java (hunk 20 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:345]: isn't `fe.size < fe.keyLen + fe.dataLen + fe.metadataLen + fe.internalMetadataLen` always true for corrupted entries?

Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:345]: Yeah, which is why we want to skip that byte :)

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java (hunk 19 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:282]: And we are 100% sure you can only have internal metadata with optimistic or async xsite?

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java (hunk 20 above)
Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:345]: > isn't `fe.size < fe.keyLen + fe.dataLen + fe.metadataLen + fe.internalMetadataLen` always true for corrupted entries?

I don't think so if the corrupt entry was subsqeuently "freed" and `fe.keyLen = 0`.

Comment (pruivo) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:345]: but in that case, you don't want to skip. you want to convert to the correct format

Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:345]: But I guess in that case `fe.keyLen <= 0` protects against that

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java (hunk 19 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:282]: I added the internal metadata for those 2 configurations. Not sure if somebody added it for something else :) 
in any case, the Builder is smart enough to detect if the internal metadata is null

Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:282]: That isn't the problem though. I was more referring to we will marshall the EMPTY PrivateMetadata for every object where neither is true. But it sounds like there should be no other uses of internal metadata.

**Comment by wburns** [comment]
I pushed a commit here that should fix the `SingleFileMigrateCorruptTest` test. The big thing was the header didn't write the new version, which actually meant it would migrate it every time, oops.

File: core/src/test/java/org/infinispan/persistence/file/SingleFileMigrateCorruptTest.java [hunk 22]
```diff
@@ -181,7 +204,6 @@ private void assertContent(Cache<Object, Object> cache) {
 //      cbh.newConfigurationBuilder(CACHE_NAME)
 //            .persistence()
 //            .addSingleFileStore()
-//            .purgeOnStartup(true)
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/persistence/file/SingleFileMigrateCorruptTest.java]: :see_no_evil: That was never going to help!

**Comment by ryanemerson** [comment]
> Also purgeOnStartup for the 12.1.bad version was set, which obviously doesn't work :)

:see_no_evil: That was never going to help! :smile: 

Thanks for looking into this @wburns and good observation about expiration. Just adding a commit to fix checkstyle and get CI running.

File: core/src/test/java/org/infinispan/persistence/file/SFSCorruptor.java [hunk 23]
```diff
@@ -0,0 +1,13 @@
+package org.infinispan.persistence.file;
+
+/**
+ * // TODO: Document this
+ *
+ * @author Ryan Emerson
+ * @since 13.0
+ */
+public class SFSCorruptor {
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/persistence/file/SFSCorruptor.java:9]: is this class supposed to be here? :)

**Review by pruivo — APPROVED** [review]

File: core/src/test/java/org/infinispan/persistence/file/SFSCorruptor.java (hunk 23 above)
Comment (ryanemerson) [reply core/src/test/java/org/infinispan/persistence/file/SFSCorruptor.java]: Whoops! Removed.

**Comment by ryanemerson** [comment]
Rebased. I've added you both as co-authors on the commit.

File: core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java [hunk 24]
```diff
@@ -297,61 +484,38 @@ private void migrateFromV11() {
...
 
-            // Read and write the remaining data from old file
-            buf = readChannelUpdateOffset(buf, offset, remainingLength);
-            newFilePos += newChannel.write(buf, newFilePos);
+            PrivateMetadata internalMeta = null;
+            if (oldFe.internalMetadataLen > 0) {
+               internalMeta = unmarshallObject(bufRef, offset, oldFe.internalMetadataLen, wrapperMissing);
+               buf = bufRef.get();
+            }
+            MarshallableEntry<? extends K, ? extends V> me = (MarshallableEntry<? extends K, ? extends V>) ctx.getMarshallableEntryFactory()
```
Comment (wburns) [inline core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:513]: We may want to add the same code as https://github.com/infinispan/infinispan/pull/9409/files#diff-3724b4c53715346242997a7b54b4c3a863ae350dd5f3911a9fb2ce0f951db398R421 so we don't write an expired entry to the file or index.

This shouldn't affect correctness though as the entry will just be ignored later when it is found to be expired, since we have the correct created and lastUsed times.

**Review by wburns — APPROVED** [review]

**Comment by wburns** [comment]
Integrated into 12.1.x, thanks @ryanemerson and @pruivo !
